### ISSUE TITLE
unable to load custom python environment with python backend 

### ISSUE THREAD

**Issue by stellaywu** [triton-inference-server/server#3480]
I'm trying to use a custom environment for a pytorch model served with the python backend 
this is the config file 
```
name: "model1"
backend: "python"

input [
  {
    name: "INPUT0"
    data_type: TYPE_FP32
    dims: [ 3 ]
  }
]
output [
  {
    name: "OUTPUT0"
    data_type: TYPE_FP32
    dims: [ 2 ]
  }


instance_group [{ kind: KIND_CPU }]

parameters: {
  key: "EXECUTION_ENV_PATH",
  value: {string_value: "$$TRITON_MODEL_DIRECTORY/model1/python-3-8.tar.gz"}
}

```

The file structure is like this 
```models
|-- model1
|   |-- 1
|   |    -- model_ckpt.pb
|   |   `-- model.py
|   |-- config.pbtxt
|   |-- python-3-8.tar.gz
|   `-- triton_python_backend_stub
```

I'm getting the error 
```UNAVAILABLE: Internal: Failed to get the canonical path for $$TRITON_MODEL_DIRECTORY/model1/python-3-8.tar.gz.```

Please help !

**Comment by Tabrizian** [comment]
What is the Triton version that you are using? `$$TRITON_MODEL_DIRECTORY` was introduced in 21.09.

**Comment by stellaywu** [comment]
Thank you @Tabrizian ! By updating the triton version to 21.09 has solved the original problem, but now I'm encountered another problem that the server doesn't start . Below are the details of the setup and error, appreciate your help! 

we are trying to deploy the 21.09 version using Kubernetes - however when pulling the image and trying to run it we get the following issues:
```
- command:
 - /bin/bash
 - -c
 args:
 - triton-server gs://bucket/model-path
 ```
The error for the above command is:

`/bin/bash: triton-server: command not found `

When we try to run it using the following command:
```
- command:
 - /bin/bash
 - -c
 args:
 - /opt/tritonserver/bin/tritonserver gs://bucket/model-path
 ```
Instead of running the server it prints out:
```
Error: Failed to initialize NVML
W1019 14:21:36.711581 1 metrics.cc:213] DCGM unable to start: DCGM initialization error
I1019 14:21:37.316686 1 libtorch.cc:1030] TRITONBACKEND_Initialize: pytorch
I1019 14:21:37.316761 1 libtorch.cc:1040] Triton TRITONBACKEND API version: 1.5
I1019 14:21:37.316781 1 libtorch.cc:1046] 'pytorch' TRITONBACKEND API version: 1.5
2021-10-19 14:21:37.599704: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
I1019 14:21:37.704036 1 tensorflow.cc:2170] TRITONBACKEND_Initialize: tensorflow
I1019 14:21:37.704099 1 tensorflow.cc:2180] Triton TRITONBACKEND API version: 1.5
I1019 14:21:37.704118 1 tensorflow.cc:2186] 'tensorflow' TRITONBACKEND API version: 1.5
I1019 14:21:37.704136 1 tensorflow.cc:2210] backend configuration:
{}
I1019 14:21:37.709603 1 onnxruntime.cc:1997] TRITONBACKEND_Initialize: onnxruntime
I1019 14:21:37.709657 1 onnxruntime.cc:2007] Triton TRITONBACKEND API version: 1.5
I1019 14:21:37.709695 1 onnxruntime.cc:2013] 'onnxruntime' TRITONBACKEND API version: 1.5
I1019 14:21:37.752110 1 openvino.cc:1193] TRITONBACKEND_Initialize: openvino
I1019 14:21:37.752172 1 openvino.cc:1203] Triton TRITONBACKEND API version: 1.5
I1019 14:21:37.752196 1 openvino.cc:1209] 'openvino' TRITONBACKEND API version: 1.5
W1019 14:21:37.752478 1 pinned_memory_manager.cc:236] Unable to allocate pinned system memory, pinned memory pool will not be available: CUDA driver version is insufficient for CUDA runtime version
I1019 14:21:37.752620 1 cuda_memory_manager.cc:115] CUDA memory pool disabled
I1019 14:21:39.417718 1 model_repository_manager.cc:1022] loading: detectron2:1
I1019 14:21:43.432737 1 python.cc:1529] Using Python execution env /tmp/folder5yvdog/python-3-8.tar.gz
I1019 14:21:43.432918 1 python.cc:1796] TRITONBACKEND_ModelInstanceInitialize: detectron2_0 (CPU device 0)
/opt/tritonserver
total 2984
-rw-rw-r-- 1 triton-server triton-server  1485 Sep 23 17:32 LICENSE
-rw-rw-r-- 1 triton-server triton-server 3012640 Sep 23 17:32 NVIDIA_Deep_Learning_Container_License.pdf
-rw-rw-r-- 1 triton-server triton-server    7 Sep 23 17:32 TRITON_VERSION
drwxr-xr-x 1 triton-server triton-server  4096 Sep 23 18:18 backends
drwxr-xr-x 2 triton-server triton-server  4096 Sep 23 18:17 bin
drwxr-xr-x 1 triton-server triton-server  4096 Sep 23 18:17 include
drwxr-xr-x 2 triton-server triton-server  4096 Sep 23 18:17 lib
-rwxrwxr-x 1 triton-server triton-server  3982 Sep 23 17:32 nvidia_entrypoint.sh
drwxr-xr-x 3 triton-server triton-server  4096 Sep 23 18:20 repoagents
drwxr-xr-x 2 triton-server triton-server  4096 Sep 23 18:17 third-party-src
```

**Comment by Tabrizian** [comment]
I think you need to wait long enough. Since your execution environment can be very large and sometimes it is required to download the environment from S3.

**Comment by stellaywu** [comment]
@Tabrizian  We've waited for a long time (hours) it still stuck at the same place, no more print.  Wondering if it has to do with the custom environment as when we remove the customized environment it went through till loading models. With the custom environment it ends at the above message.

**Comment by arshiamalek** [comment]
@Tabrizian Appreciate your help so far. We installed the custom packages using `conda` and used `conda-pack` to pack the packages and then we are serving the custom package on a cloud provider (GCP).

The `cuda` version used during the custom package installation was cuda-11-0. The version we used to create the custom environment for triton is also ` 21.09`

We then set the cuda root directory using:

`cmake -D CUDA_TOOLKIT_ROOT_DIR=/usr/local/cuda-11.0 ..`

And then we followed the steps outlined in the documentation [here](https://github.com/triton-inference-server/python_backend#using-custom-python-execution-environments).
